tss-esapi/tests/transient_key_context.rs

```diff
@@ -499,27 +432,110 @@ fn create_ecc_key_decryption_scheme() {
+            .sign(
+                key.clone(),
+                key_params,
+                auth,
+                Digest::try_from(HASH.to_vec()).unwrap(),
+            )
             .unwrap();
-        let _pub_key = ctx.read_public_key(key.clone()).unwrap();
         let _ = ctx
-            .verify_signature(key, Digest::try_from(HASH.to_vec()).unwrap(), signature)
+            .verify_signature(
+                key,
+                key_params,
+                Digest::try_from(HASH.to_vec()).unwrap(),
+                signature,
+            )
             .unwrap();
     }
 }
+
+#[test]
+fn ctx_migration_test() {
+    // Create two key contexts with a
+    let mut basic_ctx = common::create_ctx_with_session();
+    let random_digest = basic_ctx.get_random(16).unwrap();
+    let key_auth = Auth::try_from(random_digest.value().to_vec()).unwrap();
+    let prim_key_handle = basic_ctx
+        .create_primary(
+            Hierarchy::Owner,
+            &create_restricted_decryption_rsa_public(
+                SymmetricDefinitionObject::AES_256_CFB,
+                RsaKeyBits::Rsa2048,
+                RsaExponent::create(0).unwrap(),
+            )
+            .unwrap(),
+            Some(&key_auth),
+            None,
+            None,
+            None,
+        )
+        .unwrap()
+        .key_handle;
+
+    let result = basic_ctx
+        .create(
+            prim_key_handle,
+            &common::signing_key_pub(),
+            Some(&key_auth),
+            None,
+            None,
+            None,
+        )
+        .unwrap();
+
+    let key_handle = basic_ctx
+        .load(prim_key_handle, result.out_private, &result.out_public)
+        .unwrap();
+    let key_context = basic_ctx.context_save(key_handle.into()).unwrap();
+
+    let pub_key_handle = basic_ctx
+        .load_external_public(&result.out_public, Hierarchy::Owner)
+        .unwrap();
+    let pub_key_context = basic_ctx.context_save(pub_key_handle.into()).unwrap();
+
+    std::mem::drop(basic_ctx);
+
+    let mut ctx = create_ctx();
+    let key = ctx
+        .migrate_key_from_ctx(key_context, Some(key_auth.clone()))
+        .unwrap();
+    let pub_key = ctx
+        .migrate_key_from_ctx(pub_key_context, Some(key_auth.clone()))
+        .unwrap();
+
+    let key_params = KeyParams::Rsa {
+        scheme: RsaScheme::create(RsaSchemeAlgorithm::RsaSsa, Some(HashingAlgorithm::Sha256))
+            .expect("Failed to create RSA scheme"),
+        size: RsaKeyBits::Rsa2048,
+        pub_exponent: RsaExponent::default(),
+    };
+    let signature = ctx
+        .sign(
+            key,
+            key_params,
+            Some(key_auth),
+            Digest::try_from(HASH.to_vec()).unwrap(),
+        )
+        .unwrap();
+    let _ = ctx
+        .verify_signature(
+            pub_key,
+            key_params,
+            Digest::try_from(HASH.to_vec()).unwrap(),
+            signature,
+        )
+        .expect("the signature should be valid");
```

Comment:
Not sure if that's a good idea but maybe comparing the public keys before migration and after migration and making sure they're equal would be a good idea?

Currently from my reading it seems that the migration tests if the key from migration is usable but it might have been any working key that `migrate_key_from_ctx` can return. Or sign using migrated key but verify signature using non-migrated one... Maybe I'm overthinking this :thinking:

Reply:
I'll add a check. There's a couple of reasons why we don't really do this: almost every operation is integrity protected in some way, e.g. for `TPM2_Load`:
> After  inPrivate.buffer  is  decrypted  using  the  symmetric  key  of  the  parent,  the  integrity  value  shall  be 
checked before the sensitive area is used, or unmarshaled
> ...
> Before  use,  a  loaded  object  shall  be  checked  to  validate  that  the  public  and  sensitive  portions  are 
properly linked, cryptographically.

Second reason is that this tests the TPM underneath more than it tests our crate. But I think we're note very consistent on this one

Reply:
Done